From 9cc562cc126c408efb2a8220fcd67f006902412c Mon Sep 17 00:00:00 2001
From: Carlos Garnacho <carlosg@gnome.org>
Date: Thu, 15 Jan 2026 11:29:48 +0100
Subject: [PATCH 5/6] extractor: Refactor/fix handling of COMM tags

Refactor the code handling COMM tags for ID3v2.3 and v2.4, so they
share a common implementation. In doing that, this also fixes a
missing check for boundaries for ID3v2.3, which existed for v2.4.

Closes: https://gitlab.gnome.org/GNOME/localsearch/-/issues/428
---
 src/extractor/tracker-extract-mp3.c | 98 +++++++++++++----------------
 1 file changed, 44 insertions(+), 54 deletions(-)

diff --git a/src/extractor/tracker-extract-mp3.c b/src/extractor/tracker-extract-mp3.c
index ff2a2fb75..25d208b15 100644
--- a/src/extractor/tracker-extract-mp3.c
+++ b/src/extractor/tracker-extract-mp3.c
@@ -1597,6 +1597,46 @@ extract_apic_tag (id3v2tag    *tag,
 	}
 }
 
+static void
+extract_comm_tag (id3v2tag    *tag,
+                  const gchar *data,
+                  guint        pos,
+                  size_t       csize,
+                  id3tag      *info,
+                  gfloat       version)
+{
+	gchar *word = NULL;
+	gchar text_encode;
+	const gchar *text_desc;
+	const gchar *text;
+	guint offset;
+	gint text_desc_len;
+
+	text_encode =  data[pos + 0]; /* $xx */
+	text_desc = &data[pos + 4]; /* <text string according to encoding> $00 (00) */
+	text_desc_len = id3v2_strlen (text_encode, text_desc, csize - 4);
+
+	offset = 4 + text_desc_len + id3v2_nul_size (text_encode);
+
+	if (offset >= csize)
+		return;
+
+	text = &data[pos + offset]; /* <full text string according to encoding> */
+
+	if (version == 2.3f)
+		word = id3v2_text_to_utf8 (text_encode, text, csize - offset, info);
+	else
+		word = id3v24_text_to_utf8 (text_encode, text, csize - offset, info);
+
+	if (!tracker_is_empty_string (word)) {
+		g_strstrip (word);
+		g_free (tag->comment);
+		tag->comment = word;
+	} else {
+		g_free (word);
+	}
+}
+
 static void
 get_id3v24_tags (id3v24frame           frame,
                  const gchar          *data,
@@ -1615,35 +1655,9 @@ get_id3v24_tags (id3v24frame           frame,
 		extract_apic_tag (tag, data, pos, csize, info, filedata, 2.4f);
 		break;
 
-	case ID3V24_COMM: {
-		gchar *word;
-		gchar text_encode;
-		const gchar *text_desc;
-		const gchar *text;
-		guint offset;
-		gint text_desc_len;
-
-		text_encode   =  data[pos + 0]; /* $xx */
-		text_desc     = &data[pos + 4]; /* <text string according to encoding> $00 (00) */
-		text_desc_len = id3v2_strlen (text_encode, text_desc, csize - 4);
-
-		offset        = 4 + text_desc_len + id3v2_nul_size (text_encode);
-		text          = &data[pos + offset]; /* <full text string according to encoding> */
-
-		if (offset >= csize)
-			break;
-
-		word = id3v24_text_to_utf8 (text_encode, text, csize - offset, info);
-
-		if (!tracker_is_empty_string (word)) {
-			g_strstrip (word);
-			g_free (tag->comment);
-			tag->comment = word;
-		} else {
-			g_free (word);
-		}
+	case ID3V24_COMM:
+		extract_comm_tag (tag, data, pos, csize, info, 2.4f);
 		break;
-	}
 
 	case ID3V24_TMCL: {
 		extract_performers_tags (tag, data, pos, csize, info, 2.4f);
@@ -1802,33 +1816,9 @@ get_id3v23_tags (id3v24frame           frame,
 		extract_apic_tag (tag, data, pos, csize, info, filedata, 2.3f);
 		break;
 
-	case ID3V24_COMM: {
-		gchar *word;
-		gchar text_encode;
-		const gchar *text_desc;
-		const gchar *text;
-		guint offset;
-		gint text_desc_len;
-
-		text_encode   =  data[pos + 0]; /* $xx */
-		text_desc     = &data[pos + 4]; /* <text string according to encoding> $00 (00) */
-		text_desc_len = id3v2_strlen (text_encode, text_desc, csize - 4);
-
-		offset        = 4 + text_desc_len + id3v2_nul_size (text_encode);
-		text          = &data[pos + offset]; /* <full text string according to encoding> */
-
-		word = id3v2_text_to_utf8 (text_encode, text, csize - offset, info);
-
-		if (!tracker_is_empty_string (word)) {
-			g_strstrip (word);
-			g_free (tag->comment);
-			tag->comment = word;
-		} else {
-			g_free (word);
-		}
-
+	case ID3V24_COMM:
+		extract_comm_tag (tag, data, pos, csize, info, 2.4f);
 		break;
-	}
 
 	case ID3V24_IPLS: {
 		extract_performers_tags (tag, data, pos, csize, info, 2.3f);
-- 
2.52.0

